### PR TITLE
feat(token): #DRIV-67 connect to nextcloud even if token is expired

### PR DESCRIPTION
**Pull request by aymericdevriesere**
## Describe your changes
Connect to nextcloud server is working even if the user token is expired in our database. Now before every connection, we check the token validity and update it if necessary.
## Checklist tests
- [x] Delete the token on nextcloud server and try to connect from the workspace:  renew the token
- [x] Connect without deleted the token from nextcloud server:  retrieve the token in the database and connect the user.
## Issue ticket number and link
[ DRIV-67 ]
https://jira.support-ent.fr/browse/DRIV-67
## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [ ] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/java/fr/openent/nextcloud/service/impl/DefaultUserService.java [hunk 1]
```diff
@@ -91,6 +95,32 @@ private Future<Void> resolveUserSession(UserNextcloud.RequestBody userBody, User
...
+        if (userSession.isEmpty()) {
+            promise.complete(new UserNextcloud.TokenProvider());
+        } else {
+            documentsService.parametrizedListFiles(userSession, null, response -> {
+                if (response.failed()) {
+                    String messageToFormat = "[Nextcloud@%s::checkSessionValidity] Error during token validity check : %s";
+                    PromiseHelper.reject(log, messageToFormat, this.getClass().getSimpleName(), response.cause(), promise);
+                } else {
+                    if (response.result().statusCode() != 200 && response.result().statusCode() != 207) {
+                        promise.complete(new UserNextcloud.TokenProvider());
```
Comment (BastienBenard) [inline src/main/java/fr/openent/nextcloud/service/impl/DefaultUserService.java:114]: Si tu as une 4XX ou une 5XX tu renvois quand même un succès ? Ce serait pas plus adéquat de faire un promise.fail ?

**Review by sjollois — APPROVED** [review]

**Review by sjollois — APPROVED** [review]

**Review by BastienBenard — APPROVED** [review]

File: src/main/java/fr/openent/nextcloud/service/DocumentsService.java [hunk 2]
```diff
@@ -22,6 +24,16 @@ public interface DocumentsService {
...
 
+    /**
+     * Call the list API with specified handler
+     *
+     * @param userSession   User Session {@link UserNextcloud.TokenProvider}
+     * @param path          path of nextcloud's user
+     * @param handler       Specified handler
+     * @return
+     */
+    void parametrizedListFiles(UserNextcloud.TokenProvider userSession, String path, Handler<AsyncResult<HttpResponse<String>>> handler);
```
Comment (Matiyeu) [inline src/main/java/fr/openent/nextcloud/service/DocumentsService.java:35]: typo: parameterized *

File: src/main/java/fr/openent/nextcloud/service/impl/DefaultUserService.java [hunk 3]
```diff
@@ -91,6 +95,46 @@ private Future<Void> resolveUserSession(UserNextcloud.RequestBody userBody, User
...
+     * Complete the promise depending on the status code of the token check request.
+     * @param userSession   User session.
+     * @param response      Check token request (can be anything, just to check user access to his nc account).
+     * @param promise       Promise to complete.
+     */
+    private void handleSessionCheckStatus(UserNextcloud.TokenProvider userSession, AsyncResult<HttpResponse<String>> response, Promise<UserNextcloud.TokenProvider> promise) {
+        int statusCode = response.result().statusCode();
+        if (statusCode == 401) {
+            promise.complete(new UserNextcloud.TokenProvider());
+        } else if (statusCode == 200 || statusCode == 207){
```
Comment (Matiyeu) [inline src/main/java/fr/openent/nextcloud/service/impl/DefaultUserService.java:130]:
```suggestion
        } else if (statusCode == 200 || statusCode == 207) {
```

File: src/test/java/fr/openent/nextcloud/service/impl/UserServiceTest.java [hunk 4]
```diff
@@ -24,7 +26,7 @@ public class UserServiceTest {
     @Before
     public void setUp() {
         vertx = Vertx.vertx();
-        ServiceFactory serviceFactory = new ServiceFactory(null, null,null,null,null,null,null);
+        ServiceFactory serviceFactory = new ServiceFactory(vertx, null,null,null,null,null,null);
```
Comment (Matiyeu) [inline src/test/java/fr/openent/nextcloud/service/impl/UserServiceTest.java:29]:
```suggestion
        ServiceFactory serviceFactory = new ServiceFactory(vertx, null, null, null, null, null, null);
```

**Review by Matiyeu — CHANGES_REQUESTED** [review]

**Review by JinDouang — APPROVED** [review]

**Review by Matiyeu — APPROVED** [review]

**Pull request merged**